man/vipsthumbnail.1
.TH VIPSTHUMBNAIL 1 "13 May 2010"
.SH NAME
vipsthumbnail \- make thumbnails of image files
.SH SYNOPSIS
.B vipsthumbnail [flags] imagefile1 imagefile2 ...
.SH DESCRIPTION
.B vipsthumbnail(1)
processes each
.B imagefile
in turn, shrinking each image to fit within a 128 by 128 pixel square.
The shrunk image is written to a new file named
.B tn_imagefile.jpg.
This program is typically faster and uses less memory than
other image thumbnail programs.

For example:

 $ vipsthumbnail fred.png jim.tif

will read image files
.B fred.png
and
.B jim.tif
and write thumbnails to the files
.B tn_fred.jpg
and
.B tn_jim.jpg.

 $ vipsthumbnail --size=64 -o thumbnails/%s.png fred.jpg

will read image file
.B fred.jpg
and write a 64 x 64 pixel thumbnail to the file
.B thumbnails/fred.png.

.SH OPTIONS
.TP
.B -s N, --size=N
Set the output thumbnail size to
.B N
x
.B N
pixels.

You can use "MxN" to specify a rectangular bounding box.
The image is shrunk so that it just fits within this area, images
which are smaller than this are expanded.

Use "xN" or "Mx" to just resize on
one axis.

Append "<" to only resize if the input image is smaller than the
target, append ">" to only resize if the input image is larger than the target.

Append "@" to size within that many pixels.

.TP
.B --path=FORMAT
Set the output path format. Three substitutions are performed:
.B %s
is replaced by the input basename with any suffix removed,
.B %d
is replaced by the input dirname, and
.B %c
is replaced by the current working directory.

The default value is
.B %d/tn_%s.jpg
meaning JPEG output, to the same directory as the input file, with
.B tn_
prepended. You can add format options too, for example
.B %c/%s/tn_%s.jpg[Q=20]
will write JPEG images to a tree within the current directory with Q set to 20.

.TP
.B -e PROFILE, --eprofile=PROFILE
Export thumbnails with this ICC profile. Images are only colour-transformed if
there is both an output and an input profile available. The input profile can
either be embedded in the input image or supplied with the
.B --iprofile
option.

.TP
.B -i PROFILE, --iprofile=PROFILE
Import images with this ICC profile, if no profile is embedded in the image.
Images are only colour-transformed if
there is both an output and an input profile available. The output profile
should be supplied with the
.B --oprofile
option.

.TP
.B -c, --crop
Crop the output image down. The image is shrunk so as to completely fill the
bounding box in both axes, then any excess is cropped off.

.TP
.B -d, --delete
Delete the output profile from the image. This can save a small amount of
space.

.TP
.B -t, --rotate
Auto-rotate images using EXIF orientation tags.

.TP
.B -a, --linear
Shrink images in linear light colour space. This can be much slower.

.SH RETURN VALUE
returns 0 on success and non-zero on error. Error can mean one or more
conversions failed.

.SH SEE ALSO
vipsheader(1)


NEWS
VIPS changed from 7.10 to 7.12 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- IMAGE->Hist is deprecated, use im_history_get() instead
- im_append_Hist() removed
- IMAGE->Bbits is deprecated (now ignored)
- im_region_local() replaced by im_region_buffer()

VIPS enhancements
=================

- new backwards and forwards compatible VIPS file format: it has a new 
  metadata system that efficiently shares and propogates ICC profiles,
  EXIF data, etc. & whatever
- speed ups: new pixel buffer cache saves and reuses computations, uses liboil 
  where possible for a small speedup
- SMP scaling: simplified lock system improves SMP scaling, double-buffered 
  image writes overlap computation 
- lower mem use: new mmap window cache shares file mappings, history buffers 
  share and reuse image history metadata
- built-in support for new image formats: OpenEXR, Analyze7, CSV
- RGB16 and GREY16 image hints
- uses GOption for much better command-line argument handling
- better C++ binding: namespaces, VError derives from std::exception, image
  and number vector constants
- python binding 
- gcc sentinel attributes added
- added GValue as an arg type
- added im_benchmark() and friends for testing
- new functions: 
	im_analyze2vips(), im_benchmark*(), im_buildlut(),
	im_concurrency_get/set(), im_contrast_surface(),
	im_contrast_surface_raw(), im_copy_from(), im_copy_morph(),
	im_cp_desc_array(), im_cp_descv(), im_extract_areabands(),
	im_flood_blob_copy(), im_get_option_group(), im_grid(),
	im_header_exists(), im_header_map(), im_history_get(),
	im_invalidate(), im_isscalar(), im_lineset(), im_linreg(),
	im_meta*(), im_msb(), im_norm_dmask(), im_project(),
	im_raw2vips(), IM_REGION_ADDR_TOPLEFT(), im_tile_cache(),
	im_video_test()


VIPS changed from 7.8 to 7.10 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- no longer generates vips-config script, instead it makes .pc files for
  pkg-config ... replace any "vips-config" with "pkg-config vips-7.10"
- origin hint semantics changed ... it now records the position in the output
  image of the origin of the input image
- all float to int casts on pixel values are now floor() rather than rint(),
  for consistency with C casting semantics

VIPS enhancements
=================

- nip reworked for gtk+-2.4, now called nip2
- now based on glib, so VIPS no longer has it's own thread, plugin and data
  structure libraries
- i18n support, although not quite complete yet
- new error message API to make i18n possible (compatibility macros mean the
  old API still works)
- new 'start up VIPS' call, im_init_world(), sets up i18n and threads
  (optional to call it, but i18n won't fully work unless you do)
- C programs now just need "#include <vips/vips.h>", no other includes
  required
- wraps the Pango font rendering library, see im_text()
- new function flag: NOCACHE ... means the result of the call should not be
  cached, useful for video grab functions and paintbox functions
- im_embed() now supports a wider range of embedding styles, including a fast
  edge stretch
- all area operations use the new embed style to expand their input rather 
  than their output, so there are no more black borders on filtered images
- other new functions: im_render(), im_cache(), im_extract_bands(), 
  im_copy_swap(), im_rint(), im_make_xy(), im_init_world(), im_error(),
  im_warn(), im_diag(), im_rank_image(), im_replicate()
- many fixes and improvements to old funcs
- configure and portability improvements

Bug fixes
=========

- all serious bug fixes got back-ported to 7.8, I think


VIPS changes from 7.6 to 7.8 (again, not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- output format options specified in filenames: new prototypes for load and
  save tiff and jpeg
- C++ API name changes ... error => VError
- include path change
    - C includes <vips/vips.h>
    - C++ includes <vips/vipscpp.h>
- im_extract*() numbers bands from zero, not from 1 ... also other funcs which
  refer to bands (eg. im_lrmosaic() etc.)
- mosaic funcs have extra max blend width parameter

VIPS enhancements
=================

- rewritten and much fancier ip (see separate docs), now called nip ... old ip
  (slightly fixed) still there
- mmap windows remove 2GB image size limit
- images have an origin hint field
- icc transforms supported via wrapping of Little CMS
- much, much faster Fourier transforms via wrapping of libfwfft
- ppm/pgm/pbm read/write
- C++ API enhancements: new constructors, more operator equivalences
- new colourspace: Yxy
- support for D50 colour temperature images
- new functions: im_image(), im_system(), im_version*(), im_blend(),
  im_copy_set(), im_flood_blob(), im_icc_*(), im_open_local_array(),
  im_header_*(), im_sign(), im_ceil(), im_floor(), im_remainderconst*(),
  im_print(), im_guess_prefix(), im_remosaic(), im_invertlut(), Yxy funcs,
  color temperature funcs, im_clip2fmt(), im_blend(), im_lab_morph(),
  im_histnorm(), im_histcum(), im_video_v4l(), im_LabS2Lab(), im_Lab2LabS()
- new type: IMAGEVEC
- header is much faster
- ip/nip split to separate packages
- better vips2dj
- better cygwin support, mingw support too
- Mac OS X and Windows build support
- new set of sensible #define names, old names can be turned off with
  -DIM_NO_VIPS7_COMPAT
- many configure improvements

Bug fixes
=========

- speed ups
- fixes to implicit image format conversions 
- im_zoom() spots integer overflow
- im_LabS2LabQ() rounding on a/b slightly broken for a/b == 0
- fixes to refcounting in C++ API
- mask casts in C++ were broken




VIPS Changes for 7.4 to 7.6

Non-backwards compatible changes
================================

- im_histplot() has new (and better) rules
- im_andconst(), im_orconst() and im_eorconst() now take a
  double, not an unsigned char, as their constant argument type

tools/vipsthumbnail.c
/* VIPS thumbnailer
 *
 * 11/1/09
 *
 * 13/1/09
 * 	- decode labq and rad images
 * 	- colour management
 * 	- better handling of tiny images
 * 25/1/10
 * 	- added "--delete"
 * 6/2/10
 * 	- added "--interpolator"
 * 	- added "--nosharpen"
 * 	- better 'open' logic, test lazy flag now
 * 13/5/10
 * 	- oops hehe residual sharpen test was reversed
 * 	- and the mask coefficients were messed up
 * 26/5/10
 * 	- delete failed if there was a profile
 * 4/7/10
 * 	- oops sharpening was turning off for integer shrinks, thanks Nicolas
 * 30/7/10
 * 	- use new "rd" mode rather than our own open via disc
 * 8/2/12
 * 	- use :seq mode for png images
 * 	- shrink to a scanline cache to ensure we request pixels sequentially
 * 	  from the input
 * 13/6/12
 * 	- update the sequential stuff to the general method
 * 21/6/12
 * 	- remove "--nodelete" option, have a --delete option instead, off by
 * 	  default
 * 	- much more gentle extra sharpening
 * 13/11/12
 * 	- allow absolute paths in -o (thanks fuho)
 * 3/5/13
 * 	- add optional sharpening mask from file
 * 10/7/13
 * 	- rewrite for vips8
 * 	- handle embedded jpeg thumbnails
 * 12/11/13
 * 	- add --linear option
 * 18/12/13
 * 	- add --crop option
 * 5/3/14
 * 	- copy main image metadata to embedded thumbnails, thanks ottob
 * 6/3/14
 * 	- add --rotate flag
 * 7/3/14
 * 	- remove the embedded thumbnail reader, embedded thumbnails are too
 * 	  unlike the main image wrt. rotation / colour / etc.
 * 30/6/14
 * 	- fix interlaced thumbnail output, thanks lovell
 * 3/8/14
 * 	- box shrink less, use interpolator more, if window_size is large
 * 	  enough
 * 	- default to bicubic if available
 * 	- add an anti-alias filter between shrink and affine
 * 	- support CMYK
 * 	- use SEQ_UNBUF for a memory saving
 * 12/9/14
 * 	- try with embedded profile first, if that fails retry with fallback
 * 	  profile
 * 13/1/15
 * 	- exit with an error code if one or more conversions failed
 * 20/1/15
 * 	- rename -o as -f, keep -o as a hidden flag
 * 9/5/15
 * 	- use vips_resize() instead of our own code
 * 	- premultiply alpha
 * 30/7/15
 * 	- warn if you autorot and there's no exif support
 * 9/2/16
 * 	- add PDF --size support
 * 	- add SVG --size support
 * 28/2/16
 * 	- add webp --shrink support
 * 29/2/16
 * 	- deprecate sharpen and interpolate
 * 6/5/16
 * 	- restore BandFmt after unpremultiply
 * 23/5/16
 * 	- no need to guess max-alpha now premultiply does this for us
 * 1/8/16
 * 	- use scRGB as the working space in linear mode
 * 15/8/16
 * 	- can now remove 0.1 rounding adjustment
 * 2/11/16
 * 	- use vips_thumbnail(), most code moved there
 * 6/1/17
 * 	- fancy geometry strings
 * 	- support VipSize restrictions
 * 4/5/17
 * 	- add ! geo modifier
 * 30/8/17
 * 	- add --intent
 * 23/10/17
 * 	- --size Nx didn't work, argh ... thanks jrochkind
 * 3/2/20
 * 	- add --no-rotate
 * 	- add --import-profile / --export-profile names
 * 	- back to -o for output
 * 29/2/20
 * 	- deprecate --delete
 * 2/10/20
 * 	- support "stdin" as a magic input filename for thumbnail_source
 * 	- support ".suffix" as a magic output format for stdout write
 * 30/4/25
 *  - rename import/export profile args as input/oputput
 * 13/6/26
 *	- add @ specifier
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif /*HAVE_CONFIG_H*/
#include <glib/gi18n.h>

#include <stdio.h>
#include <math.h>
#include <string.h>
#include <stdlib.h>
#include <locale.h>
#include <ctype.h>

#include <vips/vips.h>
#include <vips/internal.h>

/* Default settings. We change the default to bicubic in main() if
 * this vips has been compiled with bicubic support.
 */

static char *thumbnail_size = "128";
static int thumbnail_width = 128;
static int thumbnail_height = 128;
static VipsSize size_restriction = VIPS_SIZE_BOTH;
static char *output_format = "tn_%s.jpg";
static char *output_path = NULL;
static char *output_profile = NULL;
static char *input_profile = NULL;
static gboolean linear_processing = FALSE;
static gboolean crop_image = FALSE;
static gboolean no_rotate_image = FALSE;
static char *smartcrop_image = NULL;
static char *thumbnail_intent = NULL;
static gboolean version = FALSE;
static int target_pixels = -1;

/* Deprecated and unused.
 */
static gboolean delete_profile = FALSE;
static gboolean nosharpen = FALSE;
static gboolean nodelete_profile = FALSE;
static gboolean verbose = FALSE;
static char *convolution_mask = NULL;
static char *interpolator = NULL;
static gboolean rotate_image = FALSE;

static GOptionEntry options[] = {
	{ "size", 's', 0,
		G_OPTION_ARG_STRING, &thumbnail_size,
		N_("shrink to SIZE or to WIDTHxHEIGHT"),
		N_("SIZE") },
	{ "path", 0, 0,
		G_OPTION_ARG_STRING, &output_path,
		N_("output to path FORMAT"),
		N_("FORMAT") },
	{ "output-profile", 0, 0,
		G_OPTION_ARG_FILENAME, &output_profile,
		N_("export with PROFILE"),
		N_("PROFILE") },
	{ "input-profile", 0, 0,
		G_OPTION_ARG_FILENAME, &input_profile,
		N_("import untagged images with PROFILE"),
		N_("PROFILE") },
	{ "linear", 'a', 0,
		G_OPTION_ARG_NONE, &linear_processing,
		N_("process in linear space"), NULL },
	{ "smartcrop", 'm', 0,
		G_OPTION_ARG_STRING, &smartcrop_image,
		N_("shrink and crop to fill SIZE using STRATEGY"),
		N_("STRATEGY") },
	{ "intent", 'n', 0,
		G_OPTION_ARG_STRING, &thumbnail_intent,
		N_("ICC transform with INTENT"),
		N_("INTENT") },
	{ "delete", 'd', G_OPTION_FLAG_HIDDEN,
		G_OPTION_ARG_NONE, &delete_profile,
		N_("(deprecated, does nothing)"), NULL },
	{ "no-rotate", 0, 0,
		G_OPTION_ARG_NONE, &no_rotate_image,
		N_("don't auto-rotate"), NULL },
	{ "version", 'v', 0, G_OPTION_ARG_NONE, &version,
		N_("print version"), NULL },

	/* All deprecated.
	 */
	{ "output", 'o', G_OPTION_FLAG_HIDDEN,
		G_OPTION_ARG_STRING, &output_format,
		N_("output to FORMAT"),
		N_("FORMAT") },
	{ "export-profile", 'e', G_OPTION_FLAG_HIDDEN,
		G_OPTION_ARG_FILENAME, &output_profile,
		N_("export with PROFILE"),
		N_("PROFILE") },
	{ "import-profile", 'i', G_OPTION_FLAG_HIDDEN,
		G_OPTION_ARG_FILENAME, &input_profile,
		N_("import untagged images with PROFILE"),
		N_("PROFILE") },
	{ "eprofile", 0, G_OPTION_FLAG_HIDDEN,
		G_OPTION_ARG_FILENAME, &output_profile,
		N_("export with PROFILE"),
		N_("PROFILE") },
	{ "iprofile", 0, G_OPTION_FLAG_HIDDEN,
		G_OPTION_ARG_FILENAME, &input_profile,
		N_("import untagged images with PROFILE"),
		N_("PROFILE") },
	{ "format", 'f', G_OPTION_FLAG_HIDDEN,
		G_OPTION_ARG_STRING, &output_format,
		N_("set output format string to FORMAT"),
		N_("FORMAT") },
	{ "rotate", 't', G_OPTION_FLAG_HIDDEN,
		G_OPTION_ARG_NONE, &rotate_image,
		N_("(deprecated, does nothing)"), NULL },
	{ "crop", 'c', G_OPTION_FLAG_HIDDEN,
		G_OPTION_ARG_NONE, &crop_image,
		N_("(deprecated, crop exactly to SIZE)"), NULL },
	{ "verbose", 'v', G_OPTION_FLAG_HIDDEN,
		G_OPTION_ARG_NONE, &verbose,
		N_("(deprecated, does nothing)"), NULL },
	{ "nodelete", 'l', G_OPTION_FLAG_HIDDEN,
		G_OPTION_ARG_NONE, &nodelete_profile,
		N_("(deprecated, does nothing)"), NULL },
	{ "nosharpen", 'n', G_OPTION_FLAG_HIDDEN,
		G_OPTION_ARG_NONE, &nosharpen,
		N_("(deprecated, does nothing)"), NULL },
	{ "interpolator", 'p', G_OPTION_FLAG_HIDDEN,
		G_OPTION_ARG_STRING, &interpolator,
		N_("(deprecated, does nothing)"), NULL },
	{ "sharpen", 'r', G_OPTION_FLAG_HIDDEN,
		G_OPTION_ARG_STRING, &convolution_mask,
		N_("(deprecated, does nothing)"), NULL },
	{ NULL }
};

/* Generate the filename of the output object from the input filename.
 *
 * For "a/b/c.jpg", use output_path to generate the output filename with
 * substitutions:
 *
 * - %c is whatever cwd is, so an absolute path, perhaps /d/e/f
 * - %d would be a/b
 * - %s would be c
 *
 * also supports the deprecated output_format option.
 */
static char *
thumbnail_output_filename(const char *filename)
{
	char *basename = g_path_get_basename(filename);
	char *dirname = g_path_get_dirname(filename);
	char *cwd = g_get_current_dir();

	/* Remove the suffix from basename.
	 */
	char *p;
	if ((p = strrchr(basename, '.')))
		*p = '\0';

	char path[FILENAME_MAX];
	VipsBuf buf = VIPS_BUF_STATIC(path);

	if (output_path) {
		vips_buf_appends(&buf, output_path);
		(void) vips_buf_all(&buf);

		vips__substitutec(path, FILENAME_MAX, 's', basename);
		vips__substitutec(path, FILENAME_MAX, 'd', dirname);
		vips__substitutec(path, FILENAME_MAX, 'c', cwd);
	}
	else {
		/* output_format can be a relative path, in which case we prefix with
		 * the path from the incoming file.
		 */
		if (!g_path_is_absolute(output_format)) {
			vips_buf_appends(&buf, dirname);
			vips_buf_appends(&buf, G_DIR_SEPARATOR_S);
		}
		vips_buf_appends(&buf, output_format);
		(void) vips_buf_all(&buf);

		vips__substitute(path, FILENAME_MAX, basename);
	}

	g_free(basename);
	g_free(dirname);
	g_free(cwd);

	return g_strdup(path);
}

/* If the selected output specifier is a file format (eg. ".jpg"), return
 * that, otherwise NULL.
 */
static const char *
thumbnail_output_format(void)
{
	const char *output = output_path ? output_path : output_format;

	// careful, "./%s.jpg" is a filename, not a format
	if (vips_isprefix(".", output) &&
		!vips_isprefix("./", output))
		return output;

	return NULL;
}

/* Given (eg.) "/poop/somefile.png", write @im to the thumbnail name,
 * (eg.) "/poop/tn_somefile.jpg".
 */
static int
thumbnail_write_file(VipsObject *process, VipsImage *im, const char *filename)
{
	char *output_filename = thumbnail_output_filename(filename);

	g_info("thumbnailing %s as %s", filename, output_filename);

	if (vips_image_write_to_file(im, output_filename, NULL)) {
		g_free(output_filename);
		return -1;
	}
	g_free(output_filename);

	return 0;
}

static void
thumbnail_set_size(VipsImage *image)
{
	if (target_pixels > 0) {
		gint64 source_pixels = (gint64) image->Xsize * image->Ysize;
		double shrink_factor = sqrt((double) source_pixels / target_pixels);

		thumbnail_width =
			VIPS_CLIP(1, image->Xsize / shrink_factor, VIPS_MAX_COORD);
		thumbnail_height =
			VIPS_CLIP(1, image->Ysize / shrink_factor, VIPS_MAX_COORD);
	}
}

static int
thumbnail_process(VipsObject *process, const char *name)
{
	VipsInteresting interesting;
	VipsImage *image;
	VipsIntent intent;
	char filename[VIPS_PATH_MAX];
	char option_string[VIPS_PATH_MAX];

	interesting = VIPS_INTERESTING_NONE;
	if (crop_image)
		interesting = VIPS_INTERESTING_CENTRE;
	if (smartcrop_image) {
		int n;

		if ((n = vips_enum_from_nick("vipsthumbnail",
				 VIPS_TYPE_INTERESTING, smartcrop_image)) < 0)
			return -1;
		interesting = n;
	}

	intent = VIPS_INTENT_RELATIVE;
	if (thumbnail_intent) {
		int n;

		if ((n = vips_enum_from_nick("vipsthumbnail",
				 VIPS_TYPE_INTENT, thumbnail_intent)) < 0)
			return -1;
		intent = n;
	}

	vips__filename_split8(name, filename, option_string);
	if (strcmp(filename, "stdin") == 0) {
		VipsSource *source;

		if (!(source = vips_source_new_from_descriptor(0)))
			return -1;

		if (target_pixels > 0) {
			VipsImage *plain;
			if (!(plain = vips_image_new_from_source(source, "", NULL)))
				return -1;
			thumbnail_set_size(plain);
			VIPS_UNREF(plain);
		}

		if (vips_thumbnail_source(source, &image, thumbnail_width,
				"option-string", option_string,
				"height", thumbnail_height,
				"size", size_restriction,
				"no-rotate", no_rotate_image,
				"crop", interesting,
				"linear", linear_processing,
				"import-profile", input_profile,
				"output-profile", output_profile,
				"intent", intent,
				NULL)) {
			VIPS_UNREF(source);
			return -1;
		}
		VIPS_UNREF(source);
	}
	else {
		if (target_pixels > 0) {
			VipsImage *plain;
			if (!(plain = vips_image_new_from_file(filename, NULL)))
				return -1;
			thumbnail_set_size(plain);
			VIPS_UNREF(plain);
		}

		if (vips_thumbnail(name, &image, thumbnail_width,
				"height", thumbnail_height,
				"size", size_restriction,
				"no-rotate", no_rotate_image,
				"crop", interesting,
				"linear", linear_processing,
				"import-profile", input_profile,
				"output-profile", output_profile,
				"intent", intent,
				NULL))
			return -1;
	}

	/* If the output format is something like ".jpg", we write to stdout
	 * instead.
	 */
	const char *format;
	if ((format = thumbnail_output_format())) {
		VipsTarget *target;

		if (!(target = vips_target_new_to_descriptor(1)))
			return -1;

		if (vips_image_write_to_target(image, format, target, NULL)) {
			VIPS_UNREF(image);
			VIPS_UNREF(target);
			return -1;
		}

		VIPS_UNREF(target);
	}
	else {
		if (thumbnail_write_file(process, image, name)) {
			VIPS_UNREF(image);
			return -1;
		}
	}

	VIPS_UNREF(image);

	return 0;
}

/* Parse a geometry string and set thumbnail_width and thumbnail_height.
 */
static int
thumbnail_parse_geometry(const char *geometry)
{
	/* Geometry strings have a regex like:
	 *
	 * 	^(\\d+)? (x)? (\\d+)? ([<>])?$
	 *
	 * Sadly GRegex is 2.14 and later, and we need to work with 2.6.
	 */

	const char *p;
	gboolean had_x;

	/* w or h missing means replace with a huuuge value to prevent
	 * reduction or enlargement in that axis.
	 */
	thumbnail_width = VIPS_MAX_COORD;
	thumbnail_height = VIPS_MAX_COORD;

	p = geometry;

	/* Get the width.
	 */
	while (g_ascii_isspace(*p))
		p++;
	if (g_ascii_isdigit(*p)) {
		thumbnail_width = atoi(p);

		while (g_ascii_isdigit(*p))
			p++;
	}

	/* Get the optional 'x'.
	 */
	while (g_ascii_isspace(*p))
		p++;
	had_x = FALSE;
	if (*p == 'x') {
		p += 1;
		had_x = TRUE;
	}
	while (g_ascii_isspace(*p))
		p++;

	/* Get the height.
	 */
	if (g_ascii_isdigit(*p)) {
		thumbnail_height = atoi(p);

		while (g_ascii_isdigit(*p))
			p++;
	}

	/* Get the terminating modifier,
	 */
	while (g_ascii_isspace(*p))
		p++;
	if (*p == '<')
		size_restriction = VIPS_SIZE_UP;
	else if (*p == '>')
		size_restriction = VIPS_SIZE_DOWN;
	else if (*p == '!')
		size_restriction = VIPS_SIZE_FORCE;
	else if (*p == '@')
		target_pixels = VIPS_CLIP(1, thumbnail_width, VIPS_MAX_COORD);
	else if (*p != '\0' ||
		(thumbnail_width == VIPS_MAX_COORD &&
			thumbnail_height == VIPS_MAX_COORD)) {
		vips_error("thumbnail", "%s", _("bad geometry spec"));
		return -1;
	}

	/* If target_pixels is set, we must only have width set.
	 */
	if (target_pixels != -1 &&
		had_x) {
		vips_error("thumbnail",
			"only one dimension can be specified for size by pixel count");
		return -1;
	}

	/* If there was no 'x' we have just width. vipsthumbnail history means
	 * this is a square bounding box.
	 */
	if (!had_x)
		thumbnail_height = thumbnail_width;

	/* If force is set and one of width or height isn't set, copy from the
	 * one that is.
	 */
	if (size_restriction == VIPS_SIZE_FORCE) {
		if (thumbnail_width == VIPS_MAX_COORD)
			thumbnail_width = thumbnail_height;
		if (thumbnail_height == VIPS_MAX_COORD)
			thumbnail_height = thumbnail_width;
	}

	/* If --crop is set or force is set, both width and height must be
	 * specified, since we'll need a complete bounding box to fill.
	 */
	if (crop_image ||
		smartcrop_image ||
		size_restriction == VIPS_SIZE_FORCE)
		if (thumbnail_width == VIPS_MAX_COORD ||
			thumbnail_height == VIPS_MAX_COORD) {
			vips_error("thumbnail",
				"both width and height must be given if "
				"crop is enabled");
			return -1;
		}

	return 0;
}

int
main(int argc, char **argv)
{
	GOptionContext *context;
	GOptionGroup *main_group;
	GError *error = NULL;
	int i;
	int result;

	if (VIPS_INIT(argv[0]))
		vips_error_exit("unable to start VIPS");

#ifdef ENABLE_NLS
	textdomain(GETTEXT_PACKAGE);
#endif /* ENABLE_NLS */
	setlocale(LC_ALL, "");

	{
		char *basename;

		basename = g_path_get_basename(argv[0]);
		g_set_prgname(basename);
		g_free(basename);
	}

	/* On Windows, argv is ascii-only .. use this to get a utf-8 version of
	 * the args.
	 */
#ifdef G_OS_WIN32
	argv = g_win32_get_command_line();
#endif /*G_OS_WIN32*/

	context = g_option_context_new(_("- thumbnail generator"));

	main_group = g_option_group_new(NULL, NULL, NULL, NULL, NULL);
	g_option_group_add_entries(main_group, options);
	vips_add_option_entries(main_group);
	g_option_group_set_translation_domain(main_group, GETTEXT_PACKAGE);
	g_option_context_set_main_group(context, main_group);

#ifdef G_OS_WIN32
	if (!g_option_context_parse_strv(context, &argv, &error))
#else  /*!G_OS_WIN32*/
	if (!g_option_context_parse(context, &argc, &argv, &error))
#endif /*G_OS_WIN32*/
	{
		if (error) {
			fprintf(stderr, "%s\n", error->message);
			g_error_free(error);
		}

		vips_error_exit("try \"%s --help\"", g_get_prgname());
	}

	g_option_context_free(context);

	if (version)
		printf("vips-%s\n", vips_version_string());

	if (thumbnail_size &&
		thumbnail_parse_geometry(thumbnail_size))
		vips_error_exit(NULL);

#ifndef HAVE_EXIF
	if (rotate_image)
		g_warning("auto-rotate disabled: "
				  "libvips built without exif support");
#endif /*!HAVE_EXIF*/

	result = 0;

	for (i = 1; argv[i]; i++) {
		/* Hang resources for processing this thumbnail off @process.
		 */
		VipsObject *process = VIPS_OBJECT(vips_image_new());

		if (thumbnail_process(process, argv[i])) {
			fprintf(stderr, "%s: unable to thumbnail %s\n",
				argv[0], argv[i]);
			fprintf(stderr, "%s", vips_error_buffer());
			vips_error_clear();

			/* We had a conversion failure: return an error code
			 * when we finally exit.
			 */
			result = -1;
		}

		g_object_unref(process);
	}

	/* We don't free this on error exit, sadly.
	 */
#ifdef G_OS_WIN32
	g_strfreev(argv);
#endif /*G_OS_WIN32*/

	vips_shutdown();

	vips__win32_terminate(result);
	return result;
}
